File: backend/app/schemas/social_circles.py
# backend/app/schemas/social_circles.py

"""Social Circles API schemas.

Defines request/response schemas for the social circles endpoint.
All data is inferred from existing book/author/publisher/binder relationships.
"""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class NodeType(str, Enum):
    """Types of nodes in the social network."""

    author = "author"
    publisher = "publisher"
    binder = "binder"


class ConnectionType(str, Enum):
    """Types of connections between nodes."""

    publisher = "publisher"  # Author published by publisher
    shared_publisher = "shared_publisher"  # Two authors share a publisher
    binder = "binder"  # Author's book bound by binder


class Era(str, Enum):
    """Historical eras for categorization."""

    pre_romantic = "pre_romantic"
    romantic = "romantic"
    victorian = "victorian"
    edwardian = "edwardian"
    post_1910 = "post_1910"
    unknown = "unknown"


class SocialCircleNode(BaseModel):
    """A node in the social network graph."""

    id: str = Field(..., description="Node ID (e.g., 'author:42')")
    entity_id: int = Field(..., description="Database entity ID")
    name: str = Field(..., description="Display name")
    type: NodeType = Field(..., description="Node type")

    # Author-specific fields
    birth_year: int | None = Field(None, description="Birth year (authors)")
    death_year: int | None = Field(None, description="Death year (authors)")
    era: Era | None = Field(None, description="Historical era")
    tier: str | None = Field(None, description="Tier classification")

    # Publisher/binder-specific fields
    founded_year: int | None = Field(None, description="Year founded (publishers/binders)")
    closed_year: int | None = Field(None, description="Year closed (publishers/binders)")

    # Collection stats
    book_count: int = Field(..., description="Number of books in collection")
    book_ids: list[int] = Field(default_factory=list, description="Book IDs in collection")


class SocialCircleEdge(BaseModel):
    """An edge (connection) in the social network graph."""

    id: str = Field(..., description="Edge ID (e.g., 'e:author:42:publisher:7')")
    source: str = Field(..., description="Source node ID")
    target: str = Field(..., description="Target node ID")
    type: ConnectionType = Field(..., description="Connection type")
    strength: int = Field(
        ..., ge=2, le=10, description="Connection strength (2-10, based on shared works)"
    )
    evidence: str | None = Field(None, description="Evidence for connection")
    shared_book_ids: list[int] | None = Field(None, description="Books connecting these nodes")
    start_year: int | None = Field(None, description="Start of relationship")
    end_year: int | None = Field(None, description="End of relationship")


class SocialCirclesMeta(BaseModel):
    """Metadata about the social circles response."""

    total_books: int = Field(..., description="Total books analyzed")
    total_authors: int = Field(..., description="Total authors in graph")
    total_publishers: int = Field(..., description="Total publishers in graph")
    total_binders: int = Field(..., description="Total binders in graph")
    date_range: tuple[int, int] = Field(..., description="Min/max year range")
    generated_at: datetime = Field(..., description="When data was generated")
    truncated: bool = Field(False, description="True if data was truncated due to limits")


class SocialCirclesResponse(BaseModel):
    """Response schema for the social circles endpoint."""

    nodes: list[SocialCircleNode] = Field(..., description="Graph nodes")
    edges: list[SocialCircleEdge] = Field(..., description="Graph edges")
    meta: SocialCirclesMeta = Field(..., description="Response metadata")

File: frontend/src/types/socialCircles.ts
// frontend/src/types/socialCircles.ts

/**
 * Social Circles domain types.
 * Uses branded types for type-safe IDs.
 */

// =============================================================================
// Branded ID Types
// =============================================================================

/** Node identifier (e.g., "author:42", "publisher:7") */
export type NodeId = string & { readonly __brand: "NodeId" };

/** Edge identifier (e.g., "e:author:42:publisher:7") */
export type EdgeId = string & { readonly __brand: "EdgeId" };

/** Book ID reference */
export type BookId = number & { readonly __brand: "BookId" };

// =============================================================================
// Enums
// =============================================================================

export type NodeType = "author" | "publisher" | "binder";

export type ConnectionType = "publisher" | "shared_publisher" | "binder";

export type Era = "pre_romantic" | "romantic" | "victorian" | "edwardian" | "post_1910" | "unknown";

export type Tier = "Tier 1" | "Tier 2" | "Tier 3" | null;

export type LoadingState = "idle" | "loading" | "success" | "error";

export type LayoutMode = "force" | "circle" | "grid" | "hierarchical";

export type TimelineMode = "point" | "range";

export type PlaybackSpeed = 0.5 | 1 | 2 | 5;

// =============================================================================
// API Response Types
// =============================================================================

export interface ApiNode {
  id: NodeId;
  entity_id: number;
  name: string;
  type: NodeType;
  birth_year?: number;
  death_year?: number;
  era?: Era;
  tier?: Tier;
  founded_year?: number;
  closed_year?: number;
  book_count: number;
  book_ids: BookId[];
}

export interface ApiEdge {
  id: EdgeId;
  source: NodeId;
  target: NodeId;
  type: ConnectionType;
  strength: number;
  evidence?: string;
  shared_book_ids?: BookId[];
  start_year?: number;
  end_year?: number;
}

export interface SocialCirclesMeta {
  total_books: number;
  total_authors: number;
  total_publishers: number;
  total_binders: number;
  date_range: [number, number];
  generated_at: string;
}

export interface SocialCirclesResponse {
  nodes: ApiNode[];
  edges: ApiEdge[];
  meta: SocialCirclesMeta;
}

// =============================================================================
// State Types
// =============================================================================

export interface FilterState {
  showAuthors: boolean;
  showPublishers: boolean;
  showBinders: boolean;
  connectionTypes: ConnectionType[];
  tier1Only: boolean;
  eras: Era[];
  searchQuery: string;
}

export interface TimelineState {
  currentYear: number;
  minYear: number;
  maxYear: number;
  isPlaying: boolean;
  playbackSpeed: PlaybackSpeed;
  mode: TimelineMode;
  rangeStart?: number;
  rangeEnd?: number;
}

export interface SelectionState {
  selectedNodeId: NodeId | null;
  selectedEdgeId: EdgeId | null;
  highlightedNodeIds: Set<NodeId>;
  highlightedEdgeIds: Set<EdgeId>;
  hoveredNodeId: NodeId | null;
  hoveredEdgeId: EdgeId | null;
}

// =============================================================================
// Error Types
// =============================================================================

export type ErrorType = "network" | "parse" | "not_found" | "timeout" | "unknown";

export interface AppError {
  type: ErrorType;
  message: string;
  retryable: boolean;
}

export type Result<T, E = AppError> = { success: true; data: T } | { success: false; error: E };

// =============================================================================
// Type Guards & Validators
// =============================================================================

const NODE_ID_PATTERN = /^(author|publisher|binder):\d+$/;
const EDGE_ID_PATTERN = /^e:(author|publisher|binder):\d+:(author|publisher|binder):\d+$/;

export function isNodeId(id: string): id is NodeId {
  return NODE_ID_PATTERN.test(id);
}

export function isEdgeId(id: string): id is EdgeId {
  return EDGE_ID_PATTERN.test(id);
}

export function createNodeId(type: NodeType, entityId: number): NodeId {
  return `${type}:${entityId}` as NodeId;
}

export function createEdgeId(source: NodeId, target: NodeId): EdgeId {
  return `e:${source}:${target}` as EdgeId;
}

export function parseNodeId(id: NodeId): { type: NodeType; entityId: number } {
  const [type, entityId] = id.split(":");
  return { type: type as NodeType, entityId: parseInt(entityId, 10) };
}

// Node type guards
export function isAuthorNode(node: ApiNode): node is ApiNode & { type: "author" } {
  return node.type === "author";
}

export function isPublisherNode(node: ApiNode): node is ApiNode & { type: "publisher" } {
  return node.type === "publisher";
}

export function isBinderNode(node: ApiNode): node is ApiNode & { type: "binder" } {
  return node.type === "binder";
}

// =============================================================================
// Default Values
// =============================================================================

export const DEFAULT_FILTER_STATE: FilterState = {
  showAuthors: true,
  showPublishers: true,
  showBinders: true,
  connectionTypes: ["publisher", "shared_publisher", "binder"],
  tier1Only: false,
  eras: [],
  searchQuery: "",
};

export const DEFAULT_TIMELINE_STATE: TimelineState = {
  currentYear: 1850,
  minYear: 1780,
  maxYear: 1920,
  isPlaying: false,
  playbackSpeed: 1,
  mode: "point",
  rangeStart: undefined,
  rangeEnd: undefined,
};

export const DEFAULT_SELECTION_STATE: SelectionState = {
  selectedNodeId: null,
  selectedEdgeId: null,
  highlightedNodeIds: new Set(),
  highlightedEdgeIds: new Set(),
  hoveredNodeId: null,
  hoveredEdgeId: null,
};
